Dont check Bucket Name in Nextcloud

sparrowjack63 opened the pull request:
Hi All, 

Ceph S3 storage permit "_" and uppercase in bucket name, it's more permissive that Amazon S3 rules [1].
Actually in Nextcloud, some buckets (created from CEPH) could not be mount due to the check of the bucket name with function "isBucketDnsCompatible" (come from aws sdk).
It could be better not to check bucket name in Nextcloud, it's only necessary when you create a bucket, but that is not the case in Nextcloud.

Thanks,

[1] Look at here : 
https://tracker.ceph.com/issues/36293
http://docs.ceph.com/docs/jewel/radosgw/config-ref/ => search for "rgw relaxed s3 bucket names"

sparrowjack63 committed:
Dont check Bucket Name in Nextcloud

Signed-off-by: Sebastien Cat <sparrow.jack63@gmail.com>
Signed-off-by: Sébastien Cat <sebastien.cat@inra.fr>

rullzer commented on lib/private/Files/ObjectStore/S3ConnectionTrait.php:
```diff
@@ -99,10 +99,6 @@ protected function getConnection() {
 		$this->connection = new S3Client($options);
 
-		if (!$this->connection->isBucketDnsCompatible($this->bucket)) {
-			throw new \Exception("The configured bucket name is invalid: " . $this->bucket);
```
i would be somewhat fine with not throwning the exception. But we should log then that this bucket is not strictly S3 compliant.

sparrowjack63 replied:
You're right, it could be useful to log it. Could this suggestion be suitable ?
```
		if (!$this->connection->isBucketDnsCompatible($this->bucket)) {
                         $logger = \OC::$server->getLogger();
                         $logger->warning('Bucket "' . $this->bucket . '" This bucket name is not dns compatible, it may contain invalid characters.',
                         ['app' => 'objectstore']);
                }

```

sparrowjack63 replied:
Hi, 

Finally, as Nextcloud is creating bucket if it does not exist, I think the new created bucket should respect the DNS compatibility, nevertheless, Nextcloud should accept to mount "old created" buckets that does not respect DNS compatibility (Backward compatibility, or compatibility with CEPH).
So here a new suggestion to dot that : 
```
                if (!$this->connection->isBucketDnsCompatible($this->bucket)) {
                         $logger = \OC::$server->getLogger();
                         $logger->warning('Bucket "' . $this->bucket . '" This bucket name is not dns compatible, it may contain invalid characters.',
                         ['app' => 'objectstore']);
                }

                if (!$this->connection->doesBucketExist($this->bucket)) {
                        $logger = \OC::$server->getLogger();
                        try {
                                $logger->info('Bucket "' . $this->bucket . '" does not exist - creating it.', ['app' => 'objectstore']);
                                if (!$this->connection->isBucketDnsCompatible($this->bucket)) {
                                        throw new \Exception("The bucket will not be created because the name is not dns compatible, please correct it: " . $this->bucket);
                                }
                                $this->connection->createBucket(array(
                                        'Bucket' => $this->bucket
                                ));
                                $this->testTimeout();
                        } catch (S3Exception $e) {
                                $logger->logException($e, [
                                        'message' => 'Invalid remote storage.',
                                        'level' => ILogger::DEBUG,
                                        'app' => 'objectstore',
                                ]);
                                throw new \Exception('Creation of bucket "' . $this->bucket . '" failed. ' . $e->getMessage());
                        }
                }
```

rullzer replied:
That sounds very sane :)

could I ask you to update your PR?

sparrowjack63 replied:
It's done

sparrowjack63 committed:
Warning on S3 not compatible DNS bucket name

The new created bucket should respect the DNS compatibility, nevertheless, Nextcloud should accept to mount "old created" buckets that does not respect DNS compatibility (Backward compatibility, or compatibility with CEPH).

Signed-off-by: Sébastien Cat <sebastien.cat@inra.fr>

sparrowjack63 committed:
Correct code indent

Just correct code indent.

Signed-off-by: Sébastien Cat <sebastien.cat@inra.fr>

icewind1991 commented:
Logging a warning every time we connect to s3 will quickly flood the logs.

It should probably be logged as debug only

sparrowjack63 committed:
Bucket name - Alert in debug mode

Alert in debug mode only when trying to mount non compatible DNS bucket name (in order to not flood the logs)

Signed-off-by: Sébastien Cat <sebastien.cat@inra.fr>

sparrowjack63 commented:
You're right, it's changed to alert in debug mode.

kesselb commented:
What is the point of logging this error in debug mode only? I'm not running my production instance in debug mode and you neither.

sparrowjack63 commented:
If we keep the logging in warning mode for this point, every time a user will open a S3 bucket that is not S3 compatible, you will have a new warning entry in the log files. If you have a lot of buckets and a lot of users, it could flood the logs. Admins don't like to have useless information in the logs.

kesselb commented:
> If we keep the logging in warning mode for this point, every time a user will open a S3 bucket that is not S3 compatible, you will have a new warning entry in the log files. If you have a lot of buckets and a lot of users, it could flood the logs. Admins don't like to have useless information in the logs.

I'm with you. Why logging at all? Show a proper warning on the configuration page (then people will notice).

sparrowjack63 commented:
Good Idea, I could try it for a next pull request (or if you want), but for now, it is important to be able to mount buckets compatible with CEPH without delay.

rullzer commented:
@kesselb we can't show it on the settings page because it will do :boom: before you reach that.

rullzer commented:
@sparrowjack63 we require you to sign off your commits.

See https://github.com/nextcloud/server/pull/14254/checks?check_run_id=70263515 (for a quick howto)

sparrowjack63 commented:
@rullzer it's done

rullzer commented:
Awesome stuff @sparrowjack63 !
